features/test_runner/fastlane_test_runner_helpers/fastlane_ci_output.rb

```diff
@@ -0,0 +1,197 @@
+module FastlaneCI
+  # Implementation for the FastlaneCore::Interface abstract class
+  # this is set to the fastlane runner before executing the user's Fastfile
+  class FastlaneCIOutput < FastlaneCore::Interface
+    # the file path to the log file for the output
+    attr_accessor :file_path
+
+    # The block that's being called for each new line
+    # that should be printed out to the user
+    attr_accessor :block
+
+    def initialize(file_path: nil, block: nil)
+      raise "No file path provided" if file_path.to_s.length == 0
+      raise "No block provided" if block.nil?
+      self.file_path = file_path
+      self.block = block
+    end
+
+    def log
+      return @log if @log
+
+      @log ||= Logger.new(self.file_path)
+
+      @log.formatter = proc do |severity, datetime, progname, msg|
+        "#{format_string(datetime, severity)}#{msg}\n"
+      end
+
+      @log
+    end
+
+    def format_string(datetime = Time.now, severity = "")
+      "[#{datetime.strftime('%H:%M:%S')}]: "
+    end
+
+    #####################################################
+    # @!group Messaging: show text to the user
+    #####################################################
+
+    def error(message)
+      log.error(message.to_s.red)
+      self.block.call(
+        type: :error,
+        message: message
+      )
+    end
+
+    def important(message)
+      log.warn(message.to_s.yellow)
+      self.block.call(
+        type: :important,
+        message: message
+      )
+    end
+
+    def success(message)
+      log.info(message.to_s.green)
+      self.block.call(
+        type: :success,
+        message: message
+      )
+    end
+
+    def message(message)
+      log.info(message.to_s)
+      self.block.call(
+        type: :error,
+        message: message
+      )
+    end
+
+    def deprecated(message)
+      log.error(message.to_s.deprecated)
+      self.block.call(
+        type: :error,
+        message: message
+      )
+    end
+
+    def command(message)
+      log.info("$ #{message}".cyan)
+      self.block.call(
+        type: :command,
+        message: message
+      )
+    end
+
+    def command_output(message)
+      actual = (message.split("\r").last || "") # as clearing the line will remove the `>` and the time stamp
+      actual.split("\n").each do |msg|
+        prefix = msg.include?("▸") ? "" : "▸ "
+        log.info(prefix + "" + msg.magenta)
+      end
+    end
+
+    def verbose(message)
+      # TODO: are we gonna have a verbose mode?
+      # Proposal: we can log into 2 files, one the normal output
+      #     and one with the verbose output
```

Comment:
would the verbose file be a superset of all messages?

Reply:
I haven't thought about it enough yet, and thought we'd revisit this once we implemented the first version of storing logs. Thoughts?

Reply:
Yeah, that makes sense, should we raise an `unimplemented` exception in the mean time?

Reply:
I'd say no, because every time we'd call `UI.verbose` as part of _fastlane_ (which we do a lot), it would crash.

Reply:
I'd say as we finish things up, we're gonna go through all the open `# TODO`s anyway

Reply:
❤️ sounds great